### ISSUE TITLE
Storage feature allowed to upload malicious files.

### ISSUE THREAD

**Issue by Hexife** [pagekit/pagekit#930]
## Problem

Storage feature provided in the admin panel allows to upload two kinds of html files.

One is the SVG file included html tags and javascript, and another is html file with no extension.

## Technical Details

- Pagekit version: 1.0.16
- Webserver: Apache/2.4.18
- Database: MySQL  5.7.26-0ubuntu0.16.04.1
- PHP Version: PHP 5.6.40-5+ubuntu16.04.1+deb.sury.org+1

1. Write the malicious html file
a. svg file
```
# filename : mal.svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 96 105">

<html><head><title>malicious</title></head><body><script>alert('XSS');</script></body></html>

  <g fill="#97C024" stroke="#97C024" stroke-linejoin="round" stroke-linecap="round">
    <path d="M14,40v24M81,40v24M38,68v24M57,68v24M28,42v31h39v-31z" stroke-width="12"/>
    <path d="M32,5l5,10M64,5l-6,10 " stroke-width="2"/>
  </g>
  <path d="M22,35h51v10h-51zM22,33c0-31,51-31,51,0" fill="#97C024"/>
  <g fill="#FFF">
    <circle cx="36" cy="22" r="2"/>
    <circle cx="59" cy="22" r="2"/>
  </g>
</svg>
```
b. html file with no extension
```
# filename : mal

<html><head><title>malicious</title></head><body><script>alert('XSS');</script></body></html>
```

2. Try to upload using storage feature

![upload](https://user-images.githubusercontent.com/10476944/60160548-8d6b9180-9830-11e9-9fcf-e4aede7261fb.png)

3. Access to the uploaded file using URL. URLs are able to obtain from response of "http://[domain]/pagekit/system/finder?path=%2F&root=storage" request.

![fiddler](https://user-images.githubusercontent.com/10476944/60160546-8d6b9180-9830-11e9-8409-1c29c1b913e6.png)

![run](https://user-images.githubusercontent.com/10476944/60160609-b12ed780-9830-11e9-836e-6445844cfe83.png)

**Comment by tobbexiv** [comment]
There are already some posts on XSS like #815 with the answer as quoted below:
> An XSS leak describes the injection of untrusted code in to a trusted context by (usually) somehow bypassing access control. This is not the case for your described scenario.
>
> Pagekits posts and pages provide HTML support as a feature (like Wordpress, Joomla and many other CMSs), which implies that code injection is always possible for users with write permissions. I won't consider this a problem since write permissions should only be granted to trusted parties (users with write permissions can cause a lot of harm anyway, even besides code injection).
>
> For potential further security issues please send a mail to info@pagekit.com, so that we are able to evaluate and patch the vulnerability before it is disclosed to the public.

This works with admin panel access only. Additionally, you can prevent the upload of SVG files by removing the svg ending in the system settings. The only thing which might be an issue is the upload of files without any ending which you cannot prevent.

**Comment by Hexife** [comment]
Thank you for your reply.

It is right that this vulnerability needs admin permission. 
But I think it is not an important point, because it is the fail of content filtering and the attacker who got the admin session/account using other vulnerability can access this feature.

>Additionally, you can prevent the upload of SVG files by removing the svg ending in the system settings.

Then, how about the disabling upload of SVG files as default? SVG file has the vulnerable potential and I think it is better way that the admin of the site set the SVG as allowed extension following their needs even though the extension is dangerous.

**Comment by petrk94** [comment]
I agree with @Hexife , there should be at least a option to choose over a text input field or radio buttons, which file extensions are allowed to upload (or at least a developer code note, in which files you can adjust it manually) and that files without extension get sanitized so XSS execution should be possible.

**Comment by tobbexiv** [comment]
> But I think it is not an important point, because it is the fail of content filtering and the attacker who got the admin session/account using other vulnerability can access this feature.

Someone who has the necessary permission can write HTML / Javascript everywhere: Page content, Blog posts, Comments, ... there is no need to upload a file.

> there should be at least a option to choose over a text input field or radio buttons, which file extensions are allowed to upload

This option is already there: System -> Settings and then remove / add values in the file extensions textfield.

The only thing which is missing is the possibility to prevent the upload of files without extension. This is something which should be added.

**Comment by Hexife** [comment]
> Someone who has the necessary permission can write HTML / Javascript everywhere: Page content, Blog posts, Comments, ... there is no need to upload a file.

It is right. Evil user could use page content, blog post for triggering JavaScript or embedding HTML. 
( I'm not sure but I guess that there are less limitations than file upload feature. )

But file upload feature could use as one of the attack vector and it is the reason why I reported file upload vulnerabilities. :smile: 

> This option is already there: System -> Settings and then remove / add values in the file extensions textfield.

I confirmed the options that @tobbexiv said. 
It is the reason why I suggest appending SVG to extension blacklist as default. 
It'll be better blacklist than before.

**Comment by OS-WS** [comment]
Hi all,
Does anyone know if this issue was fixed? 
If so, when and in what commit?

Thanks in advance!!
